### DISCUSSION TITLE
Domain Driven Backend Architecture

### DISCUSSION THREAD

**Discussion by danielyogel** [trpc/trpc#1376] [Ideas]
I was wondering if anyone tried to use a granular/domain-driven way of backend architecture (while using tRPC or similar).
Something a bit like GraphQL - where you add a resolver per entity (entity may be a field or a larger entity), then the they add to each other to create the final endpoints.  
I find this more domain driven and DRY.

For example, in my backend (tRPC/Zod, Prisma ORM) - each time I add a new field to the backend - I have to add it separately to:
1. Zod's schema (Query and mutation)
2. Prisma client's "include" code (Query and mutation)
3. Prisma client's "update" code (mutation)
4. sometimes to more places in code that handle authorization and side-effects (emails and etc). (Query and mutation)

In a more domain-driven approach, I would add all of those in one place that describe the field, then some framework would build the final queries/mutations (while not losing type-safety).

I think something like [https://github.com/sledorze/morphic-ts](https://github.com/sledorze/morphic-ts) would help here.

Any comment is welcome.

**Comment by KATT** [comment]
This project might be of interest too: https://github.com/CarterGrimmeisen/zod-prisma

**Reply by KATT** [reply]
Furthermore, I don't like this sort of magic, I rarely want my data layer and my API-layer to be mapped 1:1 and I really don't like `SELECT *`  (https://github.com/prisma/prisma/issues/9353) - that said, it might be beneficial to do some lower-level data access layer for your application, but don't think it'd make it more DRY for that matter.

That 1 time out of a 10 where the "auto-CRUD" creates an unintentional security vulnerability makes the whole ordeal not worth it.

To me, it feels like you want an automatically generated GraphQL-API where you have granular permissions on a field level.

**Reply by danielyogel** [reply]
> Furthermore, I don't like this sort of magic, I rarely want my data layer and my API-layer to be mapped 1:1 and I really don't like `SELECT *` ([prisma/prisma#9353](https://github.com/prisma/prisma/issues/9353)) - that said, it might be beneficial to do some lower-level data access level for your application, but don't think it'd make it more DRY for that matter.

If I understand - I think I agree - didn't meant to map the Prisma's **DB schema** (Data) to The **API's schema** (Zod) one to one, or at all, it's a bad idea. (this is what zod-prisma lib does).

I did mean to join together the **configuration** of the Prisma's **client** include/update sections, the per-field permissions/side-effects, and the Zod Schema which are part of the of a specific tRPC endpoint - are all part of the API Layer (they only use the data-layer). Many times the input/transform/return behaviour of an API endpoint are deeply connected and repeats themselves at the field level.

So I guess that some kind of a simple util/meta-schema/class/factory, where you can configure an tRPC endpoint lifecycle behavior (input-zod,  transform & permissions logic, update/include using prisma's client) on the field level, with default 'unified' behavior, while still allowing override - may be less magic, but very maintainable and DRY.

Anyway, this are just initial thoughts, feel free to completely/partly disagree :) 

Related to GraphQL, had mixed/bad experience with it, enjoying tRPC much more, but thinks that field level architecture is something GraphQL tried to partly promote.
